src/client/producer.rs

```diff
@@ -323,64 +326,246 @@ impl ProducerClient for PartitionClient {
+                // Mutex to be dropped, so that further produce() calls can
+                // start aggregating into a new batch of writes.
+                //
+                // As a side effect, this invalidates any callers performing a
+                // linger wait + flush, preventing them from flushing this new
+                // batch.
+                self.flush(None)?;
+
+                match self.batch_builder.as_mut().unwrap().try_push(data)? {
+                    TryPush::Aggregated(handle) => handle,
+                    TryPush::NoCapacity(_) => {
+                        error!(client=?self.client, "record too large for aggregator");
+                        return Err(Error::TooLarge);
+                    }
+                }
+            }
+        };
+
+        // If this is the first writer to this batch, has_linger_waiter will be
+        // false, indicating this writer should wait for the linger time before
+        // trying to flush this batch.
+        if self.has_linger_waiter {
+            // There is an existing caller handling the linger timeout.
+            return Ok(CallerRole::JustWait(handle));
+        }
+
+        // This caller is the first writer to this batch, and it should wait for
+        // the linger time before flushing THIS batch.
+        //
+        // While the writer is waiting to flush THIS batch, it is possible
+        // another call this function results in a NoCapacity being returned by
+        // the aggregator above, which would flush THIS batch.
+        //
+        // When this happens, the writer waiting for the linger time should NOT
+        // flush the newly created batch, so it uses the flush clock as a token
+        // to avoid flushing the wrong batch when calling flush().
+        self.has_linger_waiter = true;
+        Ok(CallerRole::Linger {
+            handle,
+            flush_token: self.flush_clock,
+        })
+    }
+
+    /// Asynchronously write this batch of writes to Kafka, flushing the
+    /// underlying [`Aggregator`].
+    ///
+    /// If the caller provides a `flusher_token`, the batch flush is conditional
+    /// on the token matching. If the token does not match, the batch the caller
+    /// is attempting to flush has already been flushed, and this call is a NOP.
+    fn flush(&mut self, flusher_token: Option<usize>) -> Result<()> {
+        // If this caller is is intending to conditionally flush a specific
+        // batch, verify this BatchBuilder is the batch it is indenting to
+        // flush.
+        if let Some(token) = flusher_token {
+            if token != self.flush_clock {
+                debug!(client=?self.client, "spurious batch flush call");
+                return Ok(());
+            }
+        }
+
+        debug!(client=?self.client, "flushing batch");
+
+        // Remove the batch, temporarily swapping it for a None until a new
+        // batch is built.
```

Comment:
This is NOT panic-safe. If `batch.background_flush` which calls `aggregator.flush` panics than the entire aggregator is poisoned forever[^1]. I think we should allow the aggregator to panic and only error the affected inputs instead of all future actions.

[^1]: This BTW is the reason why the stdlib mutexes implement poisoning.

Reply:
That's right - I did consider panics here, but took the view that a panicking aggregator is an exceptional event (i.e. not an recoverable error) and should be propagated up. 

We're effectively implementing poisoning here by having the next call `take()` a `None` and exploding ourselves - I think it's reasonable to expect panics to be exceptional / fatal, but open to ideas.